analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs

```diff
@@ -25,14 +25,26 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn
     public sealed record ProgramState : IEquatable<ProgramState>
     {
         public static readonly ProgramState Empty = new();
-        private ProgramState[] toArray;
+        private readonly ImmutableDictionary<IOperation, SymbolicValue> operationValue;
+        private readonly ImmutableDictionary<ISymbol, SymbolicValue> symbolValue;
+        private readonly ImmutableDictionary<CaptureId, IOperation> captureOperation;
+        private readonly ImmutableHashSet<ISymbol> preservedSymbols;
+        private readonly ImmutableStack<ExceptionState> exceptions;
 
-        private ImmutableDictionary<IOperation, SymbolicValue> OperationValue { get; init; }     // Current SymbolicValue result of a given operation
-        private ImmutableDictionary<ISymbol, SymbolicValue> SymbolValue { get; init; }
+        private ProgramState[] toArray;
+        private int? operationValueHashCode;
+        private int? symbolValueHashCode;
+        private int? captureOperationHashCode;
+        private int? preservedSymbolsHashCode;
+        private int? exceptionsHashCode;
+        private int? hashCode;
+
+        private ImmutableDictionary<IOperation, SymbolicValue> OperationValue { get => operationValue; init => SetCachedHashCodeField(value, ref operationValue, ref operationValueHashCode); }
+        private ImmutableDictionary<ISymbol, SymbolicValue> SymbolValue { get => symbolValue; init => SetCachedHashCodeField(value, ref symbolValue, ref symbolValueHashCode); }
         private ImmutableDictionary<int, int> VisitCount { get; init; }
```

Comment:
Let's put `VisitCount` last, to make it more regular. Also in the copy constructor